README.md
# fplscrapR

This package enables those interested in [Fantasy Premier League](https://fantasy.premierleague.com) to perform detailed data analysis of the game, using the FPL's JSON API. The fplscrapR functions help R users collect and parse data from the Official Fantasy Premier League website.

## Installation

You can install the released version of fplscrapR from [GitHub](https://github.com/wiscostret/fplscrapR/) with:

``` r
remotes::install_github("wiscostret/fplscrapR")

library(fplscrapR)
```
## Examples

On the [fplscrapR website](https://wiscostret.github.io/fplscrapR/), you can find examples for alle the functions available in fplscrapR, showing you how you can easily use the package to perform simple yet interesting FPL analysis.

## News

Key changes to fplscrapR v0.2.6:

* All functions updated to align with FPL’s updated API for the 21/22 season.

* Added new functions get_player_ids(), allowing you to easily pull all player ids for combining datasets, e.g. from get_player_info() and get_player_details().

* Added access to key FPL data from the 20/21 season. There is now full coverage going back to 16/17.

* Vignettes updated

* Various bug fixes

## Questions

If you have an questions or wish to collaborate, etc., find me on [GitHub](https://github.com/wiscostret) or [Twitter](https://www.twitter.com/fplscrapR) 


NEWS.md
# News for fplscrapR 0.2.6

* All functions updated to align with FPL’s updated API for the 21/22 season.

* Added new functions get_player_ids(), allowing you to easily pull all player ids for combining datasets, e.g. from get_player_info() and get_player_details().

* Added access to key FPL data from the 20/21 season. There is now full coverage going back to 16/17.

* Vignettes updated

* Various bug fixes


tests/testthat.R
library(testthat)
library(fplscrapR)

test_check("fplscrapR")


R/fplscrapR.R
# getPlayerID

#' getPlayerID
#'
#' This function fetches the FPL ID of a player given the player's full name.
#' @param name The player's full name, as listed on the official Fantasy Premier League site (for instance: "Richarlison de Andrade", not "Richarlison". You can list multiple rounds, e.g. with c().
#' @keywords player
#' @export
#' @examples
#' getPlayerID()

getPlayerID <- function(name=NULL){
  ifelse(
    is.null(name),
    return(print("You'll need to input a player name, mate.")),
    {
      elements <- fromJSON("https://fantasy.premierleague.com/drf/elements")
      elements$playername <- paste(elements$first_name,elements$second_name)
      return(elements %>% filter(playername %in% name) %>% select(playername,id))
    }
  )
}

# getPlayerName

#' getPlayerName
#'
#' This function fetches a player's full name given FPL ID.
#' @param playerid The player's FPL ID. You can list multiple rounds, e.g. with c().
#' @keywords player
#' @export
#' @examples
#' getPlayerName()

getPlayerName <- function(playerid=NULL){
  ifelse(
    is.null(playerid),
    return(print("You'll need to input an ID, mate.")),
    {
      elements <- fromJSON("https://fantasy.premierleague.com/drf/elements")
      elements$playername <- paste(elements$first_name,elements$second_name)
      return(elements %>% filter(id %in% playerid) %>% select(playername,id))
    }
  )
}

# getPlayerInfo

#' getPlayerInfo
#'
#' This function fetches selected summary information on selected players given full player name(s).
#' @param name The player's full name, as listed on the official Fantasy Premier League site (for instance: "Richarlison de Andrade", not "Richarlison"). If blank, the function fetches all players.
#' @keywords player
#' @export
#' @examples
#' getPlayerInfo()

getPlayerInfo <- function(name=NULL){
  elements <- fromJSON("https://fantasy.premierleague.com/drf/elements")
  elements$playername <- paste(elements$first_name,elements$second_name)
  ifelse(
    is.null(name),
    return(elements),
    return(elements %>% filter(playername %in% name))
  )
}


R/entry_functions.R
# get_entry()

#' get_entry
#'
#' This function fetches detailed information for a Fantasy Premier League entry given the entry ID(s).
#' @param entryid The entry ID(s). Can be found on the FPL website under 'Gameweek history' in the URL - https://fantasy.premierleague.com/entry/XXXXXX/history.
#' @keywords entry
#' @export
#' @examples
#' get_entry(1076)

get_entry <- function(entryid = NULL){
  if(is.null(entryid)) stop("You'll need to input at least one entry ID, mate.")
  if(length(entryid) != 1) stop("One at a time, please")
   {
     entry <- jsonlite::fromJSON(paste("https://fantasy.premierleague.com/api/entry/",entryid,"/",sep=""))
     return(entry)
   }
}

# get_entry_hist()

#' get_entry_hist
#'
#' This function fetches historical statistics for a Fantasy Premier League entry given the entry ID(s).
#' @param entryid The entry ID(s). Can be found on the FPL website under 'Gameweek history' in the URL - https://fantasy.premierleague.com/entry/XXXXXX/history. You can list multiple, e.g. with c().
#' @keywords entry
#' @export
#' @examples
#' get_entry_hist(1076)

get_entry_hist <- function(entryid = NULL){
  if(is.null(entryid)) stop("You'll need to input at least one entry ID, mate.")
    {
      entryhistory2 <- data.frame()
      for (i in 1:length(entryid)){
        entry <- jsonlite::fromJSON(paste("https://fantasy.premierleague.com/api/entry/",entryid[i],"/",sep=""))
        entryhistory <- jsonlite::fromJSON(paste("https://fantasy.premierleague.com/api/entry/",entryid[i],"/history","/",sep=""))
        if(length(entryhistory$past)==0) {entryhistory$past <- data.frame(season_name="",total_points="",rank="")}
        entryhistory2 <- rbind(entryhistory2,data.frame(entryhistory$past,name=paste(entry$player_first_name,entry$player_last_name)))
      }
      return(entryhistory2)
    }
}

# get_entry_season()

#' get_entry_season
#'
#' This function fetches in-season statistics for a Fantasy Premier League entry given the entry ID(s).
#' @param entryid The entry ID(s). Can be found on the FPL website under 'Gameweek history' in the URL - https://fantasy.premierleague.com/entry/XXXXXX/history. You can list multiple, e.g. with c().
#' @keywords entry
#' @export
#' @examples
#' get_entry_season(1076)

get_entry_season <- function(entryid = NULL){
  if(is.null(entryid)) stop("You'll need to input at least one entry ID, mate.")
    {
      entryseason2 <- data.frame()
      for (i in 1:length(entryid)){
        entry <- jsonlite::fromJSON(paste("https://fantasy.premierleague.com/api/entry/",entryid[i],"/",sep=""))
        entryseason <- jsonlite::fromJSON(paste("https://fantasy.premierleague.com/api/entry/",entryid[i],"/history","/",sep=""))
        entryseason2 <- rbind(entryseason2,data.frame(entryseason$current,name=paste(entry$player_first_name,entry$player_last_name)))
      }
      return(entryseason2)
    }
}

# get_entry_picks()

#' get_entry_picks
#'
#' This function fetches player picks for a Fantasy Premier League entry given the entry ID and GW.
#' @param entryid The entry ID. Can be found on the FPL website under 'Gameweek history' in the URL - https://fantasy.premierleague.com/entry/XXXXXX/history.
#' @param gw The GW for which player picks is requested.
#' @keywords entry
#' @export
#' @examples
#'
#' get_entry_picks(1076,1)

get_entry_picks <- function(entryid = NULL, gw = NULL){
  if(is.null(entryid)) stop("You'll need to input an entry ID, mate.")
  if(is.null(gw)) stop("You'll need to input a gameweek, mate.")

  picks <- jsonlite::fromJSON(paste("https://fantasy.premierleague.com/api/entry/",entryid,"/event/",gw,"/picks","/",sep=""))

  return(picks)
}

# get_entry_player_picks()

#' get_entry_player_picks()
#'
#' This function fetches player picks for a Fantasy Premier League entry given the entry ID(s) and GW(s).
#' @param entryid The entry ID(s). Can be found on the FPL website under 'Gameweek history' in the URL - https://fantasy.premierleague.com/entry/XXXXXX/history.
#' @param gw The gameweek(s) for which player picks are requested.
#' @keywords entry
#' @export
#' @examples
#' get_entry_player_picks(1076,1)

get_entry_player_picks <- function(entryid = NULL, gw = NULL){
  if(is.null(entryid)) stop("You'll need to input an entry ID, mate.")
  if(is.null(gw)) stop("You'll need to input a gameweek, mate.")

  picks3 <- data.frame()
  for (j in 1:length(entryid)) {
    picks2 <- data.frame()
    gw_started <- get_entry(entryid[j])[["started_event"]]
    gw_thisentry <- gw[gw >= gw_started]
    for (i in 1:length(gw_thisentry)) {
      picks_list <- jsonlite::fromJSON(paste("https://fantasy.premierleague.com/api/entry/",entryid[j],"/event/",gw_thisentry[i],"/picks","/",sep=""))
      picks <- picks_list$picks

      picks$event <- gw_thisentry[i]
      picks2 <- rbind(picks2, picks)
    }
    picks2$entry <- entryid[j]
    picks3 <- rbind(picks3, picks2)
  }

  player_names <- get_player_name(unique(picks3$element))
  picks3 <- merge(player_names, picks3, by.x = "id", by.y = "element")
  names(picks3) <- c("id", "playername", "position", "multiplier", "is_captain", "is_vice_captain", "event", "entry")

  picks3 <- picks3[order(picks3$entry, picks3$event), ]

  return(picks3)
}

# get_entry_captain()

#' get_entry_captain
#'
#' This function fetches captain information for a Fantasy Premier League entry given the entry ID(s) and GW(s).
#' @param entryid The entry ID(s). Can be found on the FPL website under 'Gameweek history' in the URL - https://fantasy.premierleague.com/entry/XXXXXX/history.
#' @param gw The gameweek(s) for which captain is requested.
#' @keywords entry
#' @export
#' @examples
#' get_entry_captain(1076,1)

get_entry_captain <- function(entryid = NULL, gw = NULL){
  if(is.null(entryid)) stop("You'll need to input an entry ID, mate.")
  if(is.null(gw)) stop("You'll need to input a gameweek, mate.")

  captain3 <- data.frame()
  for (j in 1:length(entryid)) {
    captain2 <- data.frame()
    gw_started <- get_entry(entryid[j])[["started_event"]]
    gw_thisentry <- gw[gw >= gw_started]
    for (i in 1:length(gw_thisentry)) {
      picks <- jsonlite::fromJSON(paste("https://fantasy.premierleague.com/api/entry/",entryid[j],"/event/",gw_thisentry[i],"/picks","/",sep=""))
      captain <- data.frame("id" = picks$picks[picks$picks$multiplier %in% c(2, 3), "element"])
      if(nrow(captain) == 0) {
        captain <- data.frame("id" = picks$picks[picks$picks$is_captain == TRUE, "element"])
      }
      captain$event <- gw_thisentry[i]
      captain2 <- rbind(captain2, captain)
    }
    captain2$entry <- entryid[j]
    captain3 <- rbind(captain3, captain2)
  }

  player_names <- get_player_name(unique(captain3$captain))
  captain3 <- merge(captain3, player_names, by = "id")
  captain3 <- captain3[c(1, 4, 3, 2)]
  names(captain3) <- c("id", "playername", "entry", "event")
  captain3 <- captain3[order(captain3$entry,captain3$event), ]

  return(captain3)
}


R/round_functions.R
# get_round_info

#' get_round_info
#'
#' This function fetches basic summary information on FPL Gameweeks.
#' @param round The FPL Gameweek number for which basic summary information is requested, e.g. 20. You can list multiple rounds, e.g. with c(). If blank, the function fetches all Gameweeks.
#' @param season To retrieve prior season data, enter a 2-digit year corresponding to the start of the historical FPL season of interest (e.g. '18' for the 2018/2019 season).
#' @keywords round
#' @export
#' @examples
#' get_round_info(2:4)

get_round_info <- function(round = NULL, season = NULL){
  ifelse(
    !is.null(season),
    {
      events <- read.csv(paste0("https://raw.githubusercontent.com/wiscostret/histfpldata/master/getroundinfo",season,".csv"),encoding="UTF-8")
    },
    {
      events <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/bootstrap-static/")$events
    })

  ifelse(
    is.null(round),
    {
      return(events)
    },
    {
      return(events %>% dplyr::filter(id %in% round))
    }
  )
}


R/league_functions.R
# get_league()

#' get_league
#'
#' This function fetches basic information for a Fantasy Premier League mini-league given the league ID and type. In the new FPL API for 19/20, this information requires authentication - so you'll be prompted to post your FPL login and password to retrieve the data.
#' @param leagueid The league ID(s). Can be found on the FPL website.
#' @param leaguetype The league type: 'classic' or 'h2h'.
#' @keywords league
#' @export

get_league <- function(leagueid = NULL, leaguetype = "classic"){
  if(is.null(leagueid)) stop("You'll need to input a league ID, mate.")
  if(length(leagueid) != 1) stop("One league at a time, please.")
  {
    fplfetchhandle <- curl::new_handle()
    curl::handle_setform(fplfetchhandle,
      login=readline("Please enter your FPL login email: "),
      password=getPass::getPass(msg="Please enter your FPL password:"),
      redirect_uri="https://fantasy.premierleague.com/a/login",
      app="plfpl-web")
    fplfetchmemory <- curl::curl_fetch_memory("https://users.premierleague.com/accounts/login/", handle = fplfetchhandle)
  }

  if(fplfetchmemory$url != "https://fantasy.premierleague.com/a/login?state=success") stop("The authentication didn't work. You've most likely entered an incorrect FPL email and/or password.")

  {
    league <- jsonlite::fromJSON(curl::curl(paste("https://fantasy.premierleague.com/api/leagues-",leaguetype,"/",leagueid,"/standings/",sep=""), handle = fplfetchhandle))
    return(league)
  }
}

# get_league_entries()

#' get_league_entries
#'
#' This function fetches a list of entries in the standings for a Fantasy Premier League mini-league given the league ID and type. In the new FPL API for 19/20, this information requires authentication - so you'll be prompted to post your FPL login and password to retrieve the data.
#' @param leagueid The league ID(s). Can be found on the FPL website.
#' @param leaguetype The league type: 'classic' or 'h2h'.
#' @param pages The number of pages of entries to fetch (there are 50 entries on each page)
#' @keywords league
#' @export

get_league_entries <- function(leagueid = NULL, leaguetype = "classic", pages = 1){
  if(is.null(leagueid)) stop("You'll need to input a league ID, mate.")
  if(length(leagueid) != 1) stop("One league at a time, please.")
  if(pages %% 1 != 0) stop("The number of pages needs to be a whole number.")
  {
    fplfetchhandle <- curl::new_handle()
    curl::handle_setform(fplfetchhandle,
      login=readline("Please enter your FPL login email: "),
      password=getPass::getPass(msg="Please enter your FPL password:"),
      redirect_uri="https://fantasy.premierleague.com/a/login",
      app="plfpl-web")
    fplfetchmemory <- curl::curl_fetch_memory("https://users.premierleague.com/accounts/login/", handle = fplfetchhandle)
    }

  if(fplfetchmemory$url != "https://fantasy.premierleague.com/a/login?state=success") stop("The authentication didn't work. You've most likely entered an incorrect FPL email and/or password.")

  {
    entries <- data.frame()

    for (i in 1:pages){

      standings <- jsonlite::fromJSON(curl::curl(paste("https://fantasy.premierleague.com/api/leagues-",leaguetype,"/",leagueid,"/standings/?","page_standings=",i,sep=""), handle = fplfetchhandle))

      entries <- rbind(entries,standings$standings$results)

    }

    return(entries)
  }
}


R/player_functions.R
# imports
#' @import jsonlite
#' @import dplyr
#' @import curl
#' @import getPass
#' @importFrom plyr rbind.fill

## quiets concerns of R CMD check re: the .'s that appear in pipelines
if(getRversion() >= "2.15.1")  utils::globalVariables(c("."))

# get_player_id

#' get_player_id
#'
#' This function fetches the FPL ID of a player given the player's full name. If blank, the function fetches all current players.
#' @param name The player's full name, as listed on the official Fantasy Premier League site (for instance: "Richarlison de Andrade", not "Richarlison". You can list multiple, e.g. with c().
#' @keywords player
#' @export
#' @examples
#' get_player_id(name="Virgil van Dijk")

get_player_id <- function(name = NULL){
  ifelse(
    is.null(name),
      {elements <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/bootstrap-static/")$elements
      elements$playername <- paste(elements$first_name,elements$second_name)
      return(elements %>% dplyr::select(playername,id))}
    ,
    {
      elements <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/bootstrap-static/")$elements
      elements$playername <- paste(elements$first_name,elements$second_name)
      return(elements %>% dplyr::filter(playername %in% name) %>% dplyr::select(playername,id))
    }
  )
}

# get_player_ids

#' get_player_ids
#'
#' This function fetches ids for all players.
#' @param season To retrieve prior season data, enter a 2-digit year corresponding to the start of the historical FPL season of interest (e.g. '17' for the 2017/2018 season).
#' @keywords player
#' @export
#' @examples
#' get_player_ids()

get_player_ids <- function(season = NULL){
  ifelse(
    is.null(season),
    {
      elements <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/bootstrap-static/")$elements
      elements$playername <- paste(elements$first_name,elements$second_name)
    },
    {
      elements <- read.csv(paste0("https://raw.githubusercontent.com/wiscostret/histfpldata/master/getplayerinfo",season,".csv"),encoding="UTF-8")
    })
  return(elements %>% dplyr::select(id,playername))
}

# get_player_name

#' get_player_name
#'
#' This function fetches a player's full name given FPL ID. If blank, the function fetches all current players.
#' @param playerid The player's FPL ID. You can list multiple, e.g. with c().
#' @keywords player
#' @export
#' @examples
#' get_player_name(playerid=300)
#'
#' get_player_name(playerid=150:160)

get_player_name <- function(playerid = NULL){
  ifelse(
    is.null(playerid),
    {
      elements <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/bootstrap-static/")$elements
      elements$playername <- paste(elements$first_name,elements$second_name)
      return(elements %>% dplyr::select(playername,id))
    }
    ,
    {
      elements <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/bootstrap-static/")$elements
      elements$playername <- paste(elements$first_name,elements$second_name)
      return(elements %>% dplyr::filter(id %in% playerid) %>% dplyr::select(playername,id))
    }
  )
}

# get_player_info

#' get_player_info
#'
#' This function fetches selected summary information on selected players given full player name(s).
#' @param name The player's full name, as listed on the official Fantasy Premier League site (for instance: "Richarlison de Andrade", not "Richarlison"). If blank, the function fetches all players.
#' @param season To retrieve prior season data, enter a 2-digit year corresponding to the start of the historical FPL season of interest (e.g. '17' for the 2017/2018 season).
#' @keywords player
#' @export
#' @examples
#' get_player_info(name=c("Jesse Lingard","Virgil van Dijk"))

get_player_info <- function(name = NULL, season = NULL){
  ifelse(
    is.null(season),
    {
      elements <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/bootstrap-static/")$elements
      elements$playername <- paste(elements$first_name,elements$second_name)
    },
    {
      elements <- read.csv(paste0("https://raw.githubusercontent.com/wiscostret/histfpldata/master/getplayerinfo",season,".csv"),encoding="UTF-8")
    })

  ifelse(
    is.null(name),
      return(elements),
      return(elements %>% dplyr::filter(playername %in% name))
  )
}

# get_player_hist

#' get_player_hist
#'
#' This function fetches historical summary information on selected players given player ID(s).
#' @param playerid The player's ID. Can be found using get_player_id(). You can list multiple , e.g. with c(). If blank, the function fetches all current players - note that takes a while to load.
#' @keywords player
#' @export
#' @examples
#' get_player_hist(playerid=200)
#'
#' get_player_hist(playerid=get_player_id("Sead Kolasinac")$id)

get_player_hist <- function(playerid = NULL){
  elements <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/bootstrap-static/")$elements
  elements$playername <- paste(elements$first_name,elements$second_name)
  ifelse(
    is.null(playerid),
    {
      histinfo <- data.frame()
      for (i in 1:nrow(elements)){
        fplboot <- jsonlite::fromJSON(url(paste("https://fantasy.premierleague.com/api/element-summary/",elements$id[i],"/",sep="")))$history_past
        histinfo <- rbind(histinfo,data.frame(fplboot,playername=rep(elements$playername[which(elements$id==elements$id[i])],length(unique(fplboot$season_name)))))}
      return(histinfo)
    },
    {
      histinfo <- data.frame()
      for (i in 1:length(playerid)){
        fplboot <- jsonlite::fromJSON(url(paste("https://fantasy.premierleague.com/api/element-summary/",playerid[i],"/",sep="")))$history_past
        histinfo <- rbind(histinfo,data.frame(fplboot,playername=rep(elements$playername[which(elements$id==playerid[i])],length(unique(fplboot$season_name)))))}
      return(histinfo)})
}

# get_player_details

#' get_player_details
#'
#' This function fetches detailed, gameweek-by-gameweek information for the current season on selected players. If parameters left blank, the function fetches all current players - note that takes a while to load.
#' @param playerid The player's ID. Can be found using get_player_id(). You can list multiple, e.g. with c(). Won't work with prior season's data.
#' @param name Alternatively, the player's full name, as listed on the official Fantasy Premier League site (for instance: "Richarlison de Andrade", not "Richarlison"). If blank, the function fetches all players.
#' @param season To retrieve prior season data, enter a 2-digit year corresponding to the start of the historical FPL season of interest (e.g. '17' for the 2017/2018 season).
#' @keywords player
#' @export
#' @examples
#' get_player_details(300)
#'
#' get_player_details(name="Virgil van Dijk")

get_player_details <- function(playerid = NULL, name = NULL, season = NULL){
  if(!is.null(playerid) & !is.null(name)) stop("Please only supply playerid OR name, not both.")

  ifelse(
    !is.null(season),
    {
      detinfo <- read.csv(paste0("https://raw.githubusercontent.com/wiscostret/histfpldata/master/getplayerdetails",season,".csv"),encoding="UTF-8")

      ifelse(
        is.null(playerid),
        ifelse(
          is.null(name),
          return(detinfo),
          return(detinfo %>% dplyr::filter(playername %in% name))),
        return(detinfo %>% dplyr::filter(playerid %in% id))
      )
    },
    {
      elements <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/bootstrap-static/")$elements
      elements$playername <- paste(elements$first_name,elements$second_name)

      ifelse(
        is.null(playerid),
        ifelse(
          is.null(name),
          {
            detinfo <- data.frame()
            for (i in 1:nrow(elements)){
              fplboot <- jsonlite::fromJSON(url(paste("https://fantasy.premierleague.com/api/element-summary/",elements$id[i],"/",sep="")))$history
              detinfo <- plyr::rbind.fill(detinfo,data.frame(cbind(fplboot,playername=elements$playername[which(elements$id==elements$id[i])])))}
            return(detinfo)
          },
          {
            detinfo <- data.frame()
            selection <- elements %>% dplyr::filter(playername %in% name) %>% dplyr::select(id) %>% unlist() %>% as.numeric()
            for (i in 1:length(selection)){
              fplboot <- jsonlite::fromJSON(url(paste("https://fantasy.premierleague.com/api/element-summary/",selection[i],"/",sep="")))$history
              detinfo <- plyr::rbind.fill(detinfo,data.frame(cbind(fplboot,playername=elements$playername[which(elements$id==selection[i])])))}
            return(detinfo)
          }),
        {
          detinfo <- data.frame()
          for (i in 1:length(playerid)){
            fplboot <- jsonlite::fromJSON(url(paste("https://fantasy.premierleague.com/api/element-summary/",playerid[i],"/",sep="")))$history
            detinfo <- plyr::rbind.fill(detinfo,data.frame(cbind(fplboot,playername=elements$playername[which(elements$id==playerid[i])])))}
          return(detinfo)
        }
      )
    }
  )
}


R/game_functions.R
# get_game_list

#' get_game_list
#'
#' This function fetches a basic list of games.
#' @param season To retrieve prior season data, enter a 2-digit year corresponding to the start of the historical FPL season of interest (e.g. '18' for the 2018/2019 season).
#' @keywords game
#' @export
#' @examples
#' get_game_list()

get_game_list <- function(season = NULL){
  ifelse(
    !is.null(season),
    {
      gamelist <- read.csv(paste0("https://raw.githubusercontent.com/wiscostret/histfpldata/master/getgamelist",season,".csv"),encoding="UTF-8")
      return(gamelist)
    },
    {
  shorts <- get_fdr() %>% dplyr::select(short_name)
  fixtures <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/fixtures/")
  gamelist <- data.frame(
    "GW"=fixtures$event,
    "id"=fixtures$id,
    "home"=fixtures$team_h,
    "team_h"=fixtures$team_h,
    "away"=fixtures$team_a,
    "team_a"=fixtures$team_a,
    "finished"=fixtures$finished,
    "kickoff"=fixtures$kickoff_time,
    "team_h_score"=fixtures$team_h_score,
    "team_a_score"=fixtures$team_a_score) %>%
    dplyr:: mutate(home = dplyr::case_when(
      home == "1" ~ shorts[1,],
      home == "2" ~ shorts[2,],
      home == "3" ~ shorts[3,],
      home == "4" ~ shorts[4,],
      home == "5" ~ shorts[5,],
      home == "6" ~ shorts[6,],
      home == "7" ~ shorts[7,],
      home == "8" ~ shorts[8,],
      home == "9" ~ shorts[9,],
      home == "10" ~ shorts[10,],
      home == "11" ~ shorts[11,],
      home == "12" ~ shorts[12,],
      home == "13" ~ shorts[13,],
      home == "14" ~ shorts[14,],
      home == "15" ~ shorts[15,],
      home == "16" ~ shorts[16,],
      home == "17" ~ shorts[17,],
      home == "18" ~ shorts[18,],
      home == "19" ~ shorts[19,],
      home == "20" ~ shorts[20,]
    )) %>%
    dplyr:: mutate(away = dplyr::case_when(
      away == "1" ~ shorts[1,],
      away == "2" ~ shorts[2,],
      away == "3" ~ shorts[3,],
      away == "4" ~ shorts[4,],
      away == "5" ~ shorts[5,],
      away == "6" ~ shorts[6,],
      away == "7" ~ shorts[7,],
      away == "8" ~ shorts[8,],
      away == "9" ~ shorts[9,],
      away == "10" ~ shorts[10,],
      away == "11" ~ shorts[11,],
      away == "12" ~ shorts[12,],
      away == "13" ~ shorts[13,],
      away == "14" ~ shorts[14,],
      away == "15" ~ shorts[15,],
      away == "16" ~ shorts[16,],
      away == "17" ~ shorts[17,],
      away == "18" ~ shorts[18,],
      away == "19" ~ shorts[19,],
      away == "20" ~ shorts[20,]
    ))
  return(gamelist)
    }
  )
}

# get_game_stats

#' get_game_stats
#'
#' This function fetches detailed statistics for a game given the game ID. It fetches ten variables that are accessed as lists with home (h) and away (a) players: goals_scored, assists, own_goals, penalties_saved, yellow_cards, red_cards, saves, bonus, and bps.
#' @param gameid The game ID. Can be found with get_game_list().
#' @keywords game
#' @export
#' @examples
#' get_game_stats(1)

get_game_stats <- function(gameid = NULL){
  if(is.null(gameid)) stop("You'll need to input at least one game ID, mate.")
  if(length(gameid) != 1) stop("One at a time, please")
   {
     fixtures <- jsonlite::fromJSON("https://fantasy.premierleague.com/api/fixtures/")
     return((fixtures %>% dplyr::filter(id %in% gameid))$stats[[1]])
   }
}
